### DISCUSSION TITLE
Lazy load types but constructor is still running for all types

### DISCUSSION THREAD

**Discussion by raza0268** [rebing/graphql-laravel#1087] [General]
i have Lazy load types enabled but constructor is still running for all types. The field part is not running but constructor for every type is running.

**Comment by mfn** [comment]
In `config/graphql.php`, did you have a key for the name of the type / query / mutation?

If you have this:
```php
    'schemas' => [
        'default' => [
            'query' => [
                MyQuery::class,
```
it needs to create the class to figure out the name.

If you use this:
```php
    'schemas' => [
        'default' => [
            'query' => [
                'myQuery' => MyQuery::class,
```
it should not and thus not call you constructor eagerly.

**Reply by raza0268** [reply]
yes i am using exactly you mentioned later. If i hit the endpoint from graphiql then the contructor of all types does not hit but if i make a hit from front end using apollo then constructor of all types hit.

**Reply by mfn** [reply]
Can you debug where this comes from?

Could be as simple as picking a type you know is definitely not involved with query from the frontend and throw an exception in the constructor?

In that stacktrace it could help us figure out where and thus why it's instantiated.

When the schema is built, `\Rebing\GraphQL\GraphQL::addTypes` is called which instantiates all types if they **do not** have a key / name already present.

So if you did that, let's see where it happens.

**Reply by raza0268** [reply]
Sir if we have multiple schemas define in our graphql.php file so for each schema do we need to add lazyload type as we register types different for each schema to work it properly?

**Reply by raza0268** [reply]
```php
<?php

declare(strict_types=1);

namespace App\GraphQL\Types;

use GraphQL\Type\Definition\Type;
use Rebing\GraphQL\Support\Type as GraphQLType;
use Rebing\GraphQL\Support\Facades\GraphQL;
use App\Models\Locale;
use Illuminate\Support\Facades\Log;

class LocaleType extends GraphQLType
{
    protected $attributes = [
        'name' => 'Locales',
        'description' => 'A type',
        'model' => Locale::class,
    ];

    public function __construct()
    {
        log::info('i am in locale screen ');
    }

    public function fields(): array
    {
        return [
            'id' => [
                'type' => Type::nonNull(Type::int()),
                'description' => 'id of Locale',
            ],
            'name' => [
                'type' => Type::nonNull(Type::string()),
                'description' => 'Name of Locale',
            ],
            'locale' => [
                'type' => Type::nonNull(Type::string()),
                'description' => 'key of Locale',
            ], 
        ];
    }
}
```
This is my locale type that is registered in common schema as shown below
```php
'common' => [
            'query' => [
                'countries' => App\GraphQL\Admin\Queries\Common\CountryQuery::class,
                'locales' =>  App\GraphQL\Admin\Queries\Common\LocaleQuery::class,
            ],
       
            'types' => [
                'Country' => App\GraphQL\Types\CountryType::class,
                'Locales' => App\GraphQL\Types\LocaleType::class,
                
            ],
            'middleware' => [],
            // Which HTTP methods to support; must be given in UPPERCASE!
            'method' => ['GET', 'POST'],
        ],
```
and 
```php
'lazyload_types' => true,
```
but when i run cou ntry query my locale type constructor automatically run. Why?

**Reply by mfn** [reply]
> Why?

I suggest to follow my proposal in https://github.com/rebing/graphql-laravel/discussions/1087#discussioncomment-6822480 , throw an exception and check the stacktrace where it's coming from.

**Comment by mfn** [comment]
I actually had a bit a spare time this week and looked and it and can confirm there's something amiss.

From what I gathered:
1. the type loader fully works. For types. With that I mean those things you define under `App\GraphQL\Types`
2. but queries and mutations don't have lazy loading
3. but when building the schema they must be resolved currently -> upfront
4. those queries/mutations references the types, e.g. the args they accept
5. hence _those_ types need to be resolved

In my testing on a large project (hundreds of queries, mutations, types), when I tried to fully embrace the `name => class` in the config file, only half the types where not instantly resolved. I.e., only half where lazy resolved, the other the moment the schema was built.
(I'm not sure why not all but half TBH).

Seems the task would be to figure out how to make the schema queries, mutations, etc. also lazy loadable. I've no idea if that can even be done.

The underlying libraries docs at https://webonyx.github.io/graphql-php/schema-definition/#lazy-loading-of-types don't give me immediately the impression this would be possible, but maybe the example isn't reflecting the possibilities and it needs more investigation.

**Reply by mfn** [reply]
It seems it was made recently possible with https://github.com/webonyx/graphql-php/pull/1418 ? 

Lighthouse is working on it -> https://github.com/nuwave/lighthouse/pull/2433

**Reply by raza0268** [reply]
Thank you so much @mfn for your valuable time. I will surely look into the webonyx you mentioned.

**Reply by mfn** [reply]
If my analysis is correct, then rebing needs to be adapted to support this (unless you use webonxy directly, certainly possible).
